db/migrate/20190118204247_move_miq_set_memberships.rb

```diff
@@ -0,0 +1,39 @@
+require 'ancestry'
+
+class MoveMiqSetMemberships < ActiveRecord::Migration[5.0]
+  class Relationship < ActiveRecord::Base
+    has_ancestry
+  end
+
+  class MiqSetMembership < ActiveRecord::Base
+  end
+
+  def up
+    say_with_time("Converting set relationships to miq_set_memberships") do
+      sets = Relationship.where(:relationship => "membership").roots
```

Comment:
ok, so my views on this have changed.

I'm guessing there are a bunch of these.

|record|ancestry|
|---|---|
|root|nil|
|first|root|
|second|root/first|

we want to create:
```ruby
Relationship.where(:relationship => "membership").not(:ancestry => nil).each do |node|
  set_id = node.root.resource_id
  MiqSetmembership.create(
            :miq_set_id  => set_id,
            :member_type => member.resource_type,
            :member_id   => member.resource_id,
            :position    => member.ancestry.count("/") +1
  )
end
```


```ruby
Relationship.where(:relationship => "membership").not(:ancestry => nil).each do |member|
  root_id = member.ancestry.split("/").first.to_i
  set_id = Relationship.select(:resource_id).where(:relationship => "membership, :ancestry => nil, :id => root_id).resource_id
  MiqSetmembership.create(
            :miq_set_id  => set_id,
            :member_type => member.resource_type,
            :member_id   => member.resource_id
            :position    => member.ancestry.count("/") +1
  )
end
```

and good news:

|ruby|sql|
|---|---|
|ancestry.count("/")|regexp_count(ancestry,'/')|
|ancestry.split("/").first.to_i|split_part(ancestry, '/', 1)::bigint|
|m1.all.each { |m| m2.create()}|insert m2 (col1, col2) select c1,c2 from m1|

... think this will be much easier sql for everyone to read without too  much mumbo jumbo

Reply:
I think this can be queried in one shot with - 

```ruby
sets = Relationship.where(:relationship => "membership")
```

and then I think there are ancestry methods to arrange records that are already in memory, so in the end it's a single query for all of the data.

Reply:
Actually, you don't even need any built in arrangement methods, since these are trivial trees:

```ruby
sets = Relationship.where(:relationship => "membership")
set_groups = sets.group_by(&:parent_id)
roots_by_id = set_groups.delete(nil).index_by(&:id)
set_groups.transform_keys! { |k| roots_by_id[k] }
```